### PR TITLE
Introduce recursive dependency resolution

### PR DESCRIPTION
**Pull request by GGG-KILLER**

### CODE REVIEW

File: libs/curseforge.ts [hunk 1]
```diff
@@ -19,45 +18,167 @@ const versionMap = {
...
-	return "https://minecraft.curseforge.com/projects/" + id + "/files/" + match[1];
-}
+request.defaults({
+	headers: {
+		["User-Agent"]: fakeUserAgent(),
+	}
+})
 
-export async function getCurseforgeUrls(id: string, versions: string[]): Promise<{ [_: string]: string }>
+export class CurseforgeLink
```
Comment (SwadicalRag) [inline libs/curseforge.ts]: VIVA

File: libs/curseforge.ts [hunk 2]
```diff
@@ -19,45 +18,167 @@ const versionMap = {
...
+		const match = url.replace(/[\\\/]*$/, '').match(/\/([0-9]+)$/);
+		if(match === null)
+			return null;
+		return match[1];
 	}
-	return urls;
 }
 
-export function getCurseforgeFileId(url: string): string|null
+export class CurseforgeMod
```
Comment (SwadicalRag) [inline libs/curseforge.ts]: OOP

File: download-mods.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	const markdownUri      = fixedArguments[2];
+	const markdownData = await readUri(markdownUri);
+	if (markdownData == null)
+	{
+		console.error("Unable to read from " + markdownUri + "!");
+		process.exit(1);
+		return;
+	}
+
+	await downloadMods(getModTables(Document.fromString(markdownData)), minecraftVersion, serverDirectory, serverDirectory + "/glua-minecraft-tools-manifest.json", console.log);
```
Comment (SwadicalRag) [inline download-mods.ts]: im ok with this but i think this will trigger cake because it's a loooong line
how about

```js
await downloadMods(
    getModTables(Document.fromString(markdownData)),
    minecraftVersion,
    serverDirectory,
    serverDirectory + "/glua-minecraft-tools-manifest.json",
    console.log
);
```

also isnt this deprecated

File: libs/download.ts [hunk 4]
```diff
@@ -22,6 +22,12 @@ export interface IDownloadProgress {
 	}
 }
 
+interface IDownloadQueueItem {
```
Comment (SwadicalRag) [inline libs/download.ts:25]: Kuchen syntax for this usually looks like

```ts
interface IAmAnInterface {
	wow:    string;
	so:     string;
	pretty: string;
}
```

> always remember, tabs for indentation, spaces for alignment
> \- Kuchen

File: libs/download.ts [hunk 5]
```diff
@@ -165,7 +193,7 @@ export async function downloadMods(modTables: ITable[], minecraftVersion: string
 			{
 				if (e.code != "ENOENT") { throw e; }
 			}
-			log("- " + packModId(namespace, id) + " " + fileName);
+			log(" - " + packModId(namespace, id) + " " + fileName);
```
Comment (SwadicalRag) [inline libs/download.ts:189]: excellent

File: libs/curseforge.ts [hunk 6]
```diff
@@ -19,45 +15,113 @@ const versionMap = {
...
-	for (let version of versions)
+	public id: string;
+	public urls: StringMap<string>;
+
+	public availableForVersion(version: string)
+	{
+		return !!this.urls[version];
+	}
+
+	private async *resolveDependencies(version: string): AsyncIterableIterator<Mod>
```
Comment (SwadicalRag) [inline libs/curseforge.ts]: > IterableIterator

I approve
image

**Review by SwadicalRag — APPROVED** [review]

File: download-mods.ts (hunk 3 above)
Comment (GGG-KILLER) [reply download-mods.ts]: This `download-mods.ts` file? I re-added it because I honestly don't want to use Technic launcher and need something to facilitate downloading the mods.

I don't mind if this file gets removed though, I just thought others might want it too ¯\_(ツ)_/¯

File: download-mods.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+{
+	const [fixedArguments] = parseArguments(argc, argv);
+	let argumentsValid = fixedArguments.length == 3;
+	if (!argumentsValid)
+	{
+		console.error("Usage: ts-node download-mods.ts <directory> <minecraft-version> <mods.md>");
+		process.exit(1);
+	}
+
+	const serverDirectory  = fixedArguments[0];
```
Comment (notcake) [inline download-mods.ts]: minor: this is no longer a `serverDirectory` but a `modsDirectory`

File: download-mods.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
+import { Document } from "./libs/markdown";
+import { downloadMods } from "./libs/download";
+import { getModTables } from "./libs/mod-table";
+import { parseArguments, readUri } from "./libs/utils";
+
+async function main(argc: number, argv: string[])
+{
+	const [fixedArguments] = parseArguments(argc, argv);
+	let argumentsValid = fixedArguments.length == 3;
+	if (!argumentsValid)
```
Comment (notcake) [inline download-mods.ts]: minor: inline the condition please

File: libs/curseforge.ts [hunk 9]
```diff
@@ -2,14 +2,10 @@
 // a common file containing the API to interact with the curseforge site
 import * as request from "request-promise";
 const fakeUserAgent = require("fake-useragent");
+type StringMap<V> = {[_: string]: V};
```
Comment (notcake) [inline libs/curseforge.ts]: o k (this is now inconsistent with every place where we have a map)

minor: whitespace: `{ [_: string]: V }`

File: libs/curseforge.ts [hunk 10]
```diff
@@ -19,45 +15,113 @@ const versionMap = {
...
+	public availableForVersion(version: string)
+	{
+		return !!this.urls[version];
+	}
+
+	private async *resolveDependencies(version: string): AsyncIterableIterator<Mod>
 	{
-		tasks[version] = getCurseforgeLinkForVersion(id, version);
+		const body: string = await request.get(`https://minecraft.curseforge.com/projects/${this.id}/relations/dependencies?filter-related-dependencies=3`);
+		if(!body) throw new Error("Unable to resolve dependencies for " + this.id + ".");
```
Comment (notcake) [inline libs/curseforge.ts]: minor: whitespace and style in this entire function

`if (expr) { return; }`
curly braces are never omitted

File: libs/curseforge.ts [hunk 11]
```diff
@@ -19,45 +15,113 @@ const versionMap = {
...
+			// circular dependencies (I'm looking at you Tesla Core Lib)
+			if(dep !== null)
+			{
+				if(dep.availableForVersion(version))
+				{
+					yield dep;
+				}
+			}
+		}
 	}
```
Comment (notcake) [inline libs/curseforge.ts:74]: `getDependencies` should only return _direct_ dependencies

File: libs/curseforge.ts [hunk 12]
```diff
@@ -19,45 +15,113 @@ const versionMap = {
...
+		this.urls = {};
+		for (let version of versions)
+		{
+			const body = await request.get(`https://minecraft.curseforge.com/projects/${this.id}/files?filter-game-version=${versionMap[version]}`);
+			const regex = /\/projects\/[^\/]+\/files\/([0-9]+)\/download/;
+			const match = body.match(regex);
+			if (match === null){ continue; }
+			this.urls[version] = `https://minecraft.curseforge.com/projects/${this.id}/files/${match[1]}`;
+		}
+	}
```
Comment (notcake) [inline libs/curseforge.ts:127]: wat.
We do not want to fetch new URLs when downloading. The hackmd URLs are pinned to specific versions.

File: libs/curseforge.ts [hunk 13]
```diff
@@ -19,45 +15,113 @@ const versionMap = {
...
-versionMap["1.12.2"] = versionMap["1.12"];
 
-async function getCurseforgeLinkForVersion(id: string, version: string): Promise<string|null>
+request.defaults({
+	headers: {
+		["User-Agent"]: fakeUserAgent(),
+	}
+})
+
+export function getCurseforgeFileID(url: string): string|null
```
Comment (notcake) [inline libs/curseforge.ts]: why did you change the capitalization

Comment (GGG-KILLER) [reply libs/curseforge.ts]: I actually didn't restore the function, I typed it out of memory and didn't realize this, I'm sorry.

File: download-mods.ts [hunk 14]
```diff
@@ -6,14 +6,13 @@ import { parseArguments, readUri } from "./libs/utils";
...
-	let argumentsValid = fixedArguments.length == 3;
-	if (!argumentsValid)
+	if (fixedArguments.length != 3)
 	{
 		console.error("Usage: ts-node download-mods.ts <directory> <minecraft-version> <mods.md>");
 		process.exit(1);
 	}
 
-	const serverDirectory  = fixedArguments[0];
+	const modsDirectory  = fixedArguments[0];
```
Comment (notcake) [inline download-mods.ts:15]: minor: alignment whitespace

If you were to file a pull request with a single commit adding `download-mods.ts` I would accept it.

File: libs/curseforge.ts [hunk 15]
```diff
@@ -31,57 +29,61 @@ export function getCurseforgeFileID(url: string): string|null
...
 		return !!this.urls[version];
 	}
 
-	private async *resolveDependencies(version: string): AsyncIterableIterator<Mod>
+	private dependencies: { [_: string]: Mod[] } = {};
+	public async getDependencies(version: string): Promise<Mod[]>
 	{
-		const body: string = await request.get(`https://minecraft.curseforge.com/projects/${this.id}/relations/dependencies?filter-related-dependencies=3`);
-		if(!body) throw new Error("Unable to resolve dependencies for " + this.id + ".");
+		if(this.dependencies[version] === undefined)
```
Comment (notcake) [inline libs/curseforge.ts]: minor: whitespace style
`if (`
`for (`

File: libs/curseforge.ts [hunk 16]
```diff
@@ -107,7 +109,7 @@ export class Mod
 		}
 	}
 
-	private static readonly ModLUT: StringMap<Mod|null> = {};
+	private static readonly ModLUT: { [_: string]: Mod|null } = {};
 	public static async fromID(id: string, versions: string[]) : Promise<Mod|null>
 	{
 		if(this.ModLUT[id] === undefined)
```
Comment (notcake) [inline libs/curseforge.ts]: * Circular dependency resolution doesn't really belong in the "constructor" for `Mod`s.
* Neither does `versions`. What if two parts of code want to fetch different version sets?

File: download-mods.ts (hunk 14 above)
Comment (notcake) [reply download-mods.ts:15]: I don't want to accept the rest of the changes because:
* Mod download now involves another billion http requests
* Incremental download now involves a billion http requests even if the hackmd has not changed.
* Downloading the same markdown twice will no longer always produce the same downloads.

File: libs/curseforge.ts (hunk 16 above)
Comment (GGG-KILLER) [reply libs/curseforge.ts]: Removed circular dependency resolution and version management from the constructor. It only accepts the ID and existing urls now.

File: download-mods.ts (hunk 14 above)
Comment (GGG-KILLER) [reply download-mods.ts:15]: About these:

- Mod download involves a billion HTTP requests because of recursive dependency resolution, although that can be not done by removing the dependency resolution lines from `downloads.ts`
- Incremental download involves a billion HTTP requests because dependencies *can* change even if the hackmd has not
- Downloading the same markdown will produce the same downloads if the mods' dependencies haven't changed at curseforge

Comment (GGG-KILLER) [reply download-mods.ts:15]: I believe the problem of excessive HTTP requests can be solved if the file upload time at curseforge was considered and dependencies were stored in the manifest along with the download time of the mod.

That way the dependencies would only be retrieved and updated if the upload time of the file at curseforge was after the time stored on the manifest.

Or alternatively, the file "versions" could be compared and dependencies only retrieved if the version on the markdown was greater than the version on the manifest itself.
